### PR TITLE
Cam Controller Free Movement

### PR DESCRIPTION
**Pull request by shadowscion**
Adds a flag that modifies mouse command input (seems to be predicted automatically) to allow 360 degree rotation. 

An additional input which zeroes out the roll has been created for use with this feature.

### CODE REVIEW

**Comment by shadowscion** [comment]
[Example Video](https://dl.dropboxusercontent.com/s/nqn10z293fmqo5x/Garry%27s%20Mod%208_24_2018%207_45_20%20PM.mp4)

[E2 used in video](https://hastebin.com/raw/ifuyapudob)

**Comment by thegrb93** [comment]
These hooks shouldn't be added until the user is using a pod controller. Can we change that?

Also the math can be more accurate and optimized. I'll give code for that once I test it.

**Comment by thegrb93** [comment]
Here's the optimized code. From my tests, it gives the same result, but I can guarantee that it's much more efficient than using all those rotatearoundaxis

```lua
hook.Add("InputMouseApply", "wire_camera_controller_input_unlock", function(cmd, x, y, ang)
	if not enabled then return false end
	if not FreeMove then return false end
	if not IsValid( self ) then enabled = false return false end
	-- feels correct, might not be, but raw values were definitely too fast
	local smooth = mouse_sensitvity:GetFloat() * FrameTime()
	local matrix = Matrix()
	matrix:SetAngles( ang )
	-- could make this a number instead
	if unroll then
		local parent, hasParent = GetParent()
		if hasParent then
			local rolldiff = math.deg( math.asin( parent:GetUp():Dot( matrix:GetRight() ) ) )
			matrix:Rotate( Angle( y * smooth, -x * smooth, rolldiff * smooth ) )
		else
			matrix:Rotate( Angle( y * smooth, -x * smooth, -ang.r * smooth ) )
		end
	else
		matrix:Rotate( Angle( y * smooth, -x * smooth, 0 ) )
	end
	cmd:SetViewAngles( matrix:GetAngles() )
	return true
end)
```

**Comment by thegrb93** [comment]
If something doesn't feel right with this change, lemme know and I can fix it

**Comment by thegrb93** [comment]
Also I just realized that your `if hasParent then` branch will only work if the camera's forward is aligned with the parent's forward.

**Comment by thegrb93** [comment]
I think this might be right
```lua
hook.Add("InputMouseApply", "wire_camera_controller_input_unlock", function(cmd, x, y, ang)
	if not enabled then return false end
	if not FreeMove then return false end
	if not IsValid( self ) then enabled = false return false end
	-- feels correct, might not be, but raw values were definitely too fast
	local smooth = mouse_sensitvity:GetFloat() * FrameTime()
	local matrix = Matrix()
	matrix:SetAngles( ang )
	-- could make this a number instead
	if unroll then
		local parent, hasParent = GetParent()
		if hasParent then
			local parentMatrix = Matrix()
			parentMatrix:SetAngles( parent:GetAngles() )
			local diffMatrix = matrix:GetInverseTR() * parentMatrix
			matrix:Rotate( Angle( y * smooth, -x * smooth, diffMatrix:GetAngles().r * smooth ) )
		else
			matrix:Rotate( Angle( y * smooth, -x * smooth, -ang.r * smooth ) )
		end
	else
		matrix:Rotate( Angle( y * smooth, -x * smooth, 0 ) )
	end
	cmd:SetViewAngles( matrix:GetAngles() )
	return true
end)
```

**Comment by shadowscion** [comment]
Feels good, and looks a whole lot cleaner, thanks. 

There is one small issue but it might be with my test E2, as the problem is amplified when the parent is rotating slower. 

With UnRoll **1**, rotating the camera too far past the parent's angle now [results in this](https://dl.dropbox.com/s/azeh4hxhxbcy49o/Garry%27s%20Mod%208_25_2018%207_43_28%20AM.mp4).

I tested the original code and it doesn't do this, but it does experiences issues of not maintaining the proper angle, so even with the shaking yours is still better.

**Comment by Divran** [comment]
Please add a comment about the "UnRoll" wire input to the C menu checkbox description. Something like, "The input UnRoll is only used if this option is enabled" or similar.

**Comment by thegrb93** [comment]
Ye I'll look at the roll calculation again later.

**Comment by thegrb93** [comment]
I think it's because when their forward vectors oppose each other, the roll gets inverted. I can make a simple hack to fix that.

**Comment by thegrb93** [comment]
This seems to do the trick
```lua
hook.Add("InputMouseApply", "wire_camera_controller_input_unlock", function(cmd, x, y, ang)
	if not enabled then return false end
	if not FreeMove then return false end
	if not IsValid( self ) then enabled = false return false end
	-- feels correct, might not be, but raw values were definitely too fast
	local smooth = mouse_sensitvity:GetFloat() * FrameTime()
	local matrix = Matrix()
	matrix:SetAngles( ang )
	-- could make this a number instead
	if unroll then
		local parent, hasParent = GetParent()
		if hasParent then
			local parentMatrix = Matrix()
			parentMatrix:SetAngles( parent:GetAngles() )
			local diffAngles = (matrix:GetInverseTR() * parentMatrix):GetAngles()
			if math.abs(diffAngles.y) > 90 then diffAngles.r = -diffAngles.r end
			matrix:Rotate( Angle( y * smooth, -x * smooth, diffAngles.r * smooth ) )
		else
			matrix:Rotate( Angle( y * smooth, -x * smooth, -ang.r * smooth ) )
		end
	else
		matrix:Rotate( Angle( y * smooth, -x * smooth, 0 ) )
	end
	cmd:SetViewAngles( matrix:GetAngles() )
	return true
end)
```

**Comment by thegrb93** [comment]
You'll still get gimble lock when the camera pitch goes 90 degrees with the parent's pitch, but that's normal since roll is undefined at that point.

**Comment by shadowscion** [comment]
Nicely done.

And wow, that's the second typo I've made in as many commit descriptions.

File: lua/entities/gmod_wire_cameracontroller.lua [hunk 1]
```diff
@@ -199,6 +201,36 @@ if CLIENT then
 		end
 	end)
 
+	local mouse_sensitvity = GetConVar("sensitivity")
+
+	hook.Remove("InputMouseApply", "wire_camera_controller_input_unlock")
+	hook.Add("InputMouseApply", "wire_camera_controller_input_unlock", function(cmd, x, y, ang)
+		if not enabled then return false end
```
[the remark this answers is not in the document]
Comment (shadowscion) [reply lua/entities/gmod_wire_cameracontroller.lua]: I had no idea hooks work that way.

**Comment by thegrb93** [comment]
Yeah good catch. I just wrongfully assumed the hook required false for no change.

**Pull request merged**